### PR TITLE
Add REST implementation for Validator's `SubscribeCommitteeSubnets`

### PR DESCRIPTION
**Pull request by PatriceVignola**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

It adds the `SubscribeCommitteeSubnets` REST API endpoint for the validator.

**Other notes for review**

I started working on this feature before https://github.com/prysmaticlabs/prysm/pull/11723 and all the tests were passing. But after merging with this change, the E2E tests started logging errors in the beacon node (although the tests are still passing somehow). I believe this is because `StateBySlot` is returning an error if the slot is in the future, but epochs at `epoch + 1` should be allowed as per the API spec. The unit tests are still passing because `StateBySlot` is mocked inside the validator duties tests, which means that this code doesn't get executed for `epoch + 1`. I'm not familiar enough with that part of the code to make a change and be sure that it doesn't break something else, especially since the change was a bug fix for another issue in the first place. It would be ideal if this issue gets fixed by another PR before my PR gets merged.

Also, I had to slightly modify the signature of the interface in this PR since validator indices are needed to get the duties, and retrieving the duties is necessary to get the `committees_at_slot` values.

### CODE REVIEW

File: validator/client/validator.go [hunk 1]
```diff
@@ -667,14 +671,18 @@ func (v *validator) subscribeToSubnets(ctx context.Context, res *ethpb.DutiesRes
 			subscribeSlots = append(subscribeSlots, attesterSlot)
 			subscribeCommitteeIndices = append(subscribeCommitteeIndices, committeeIndex)
 			subscribeIsAggregator = append(subscribeIsAggregator, aggregator)
+			subscribeValidatorIndices = append(subscribeValidatorIndices, validatorIndex)
```
Comment (james-prysm) [inline validator/client/validator.go:674]: cool so this is now collecting indicies for both current and next epochs, I saw a check on our old gRPC that tries to collect a list of indicies based on head and their respective epoch with this I think we just collect both here.

File: validator/client/beacon-api/subscribe_committee_subnets.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+		return errors.New("committee subnets subscribe request is nil")
+	}
+
+	if len(in.CommitteeIds) != len(in.Slots) || len(in.CommitteeIds) != len(in.IsAggregator) || len(in.CommitteeIds) != len(validatorIndices) {
+		return errors.New("arrays `in.CommitteeIds`, `in.Slots`, `in.IsAggregator` and `validatorIndices` don't have the same length")
+	}
+
+	slotToCommitteesAtSlotMap := make(map[types.Slot]uint64)
+	jsonCommitteeSubscriptions := make([]*apimiddleware.BeaconCommitteeSubscribeJson, len(in.CommitteeIds))
+	for index := range in.CommitteeIds {
```
Comment (james-prysm) [inline validator/client/beacon-api/subscribe_committee_subnets.go:27]: was wondering why it was looping through the committeeIds first but saw the check `len(in.CommitteeIds) != len(in.Slots)`

**Comment by james-prysm** [comment]
I believe there was an additional fix for attester duties in #11806

File: validator/client/beacon-api/subscribe_committee_subnets_test.go [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+	"github.com/prysmaticlabs/prysm/v3/beacon-chain/rpc/apimiddleware"
+	types "github.com/prysmaticlabs/prysm/v3/consensus-types/primitives"
+	ethpb "github.com/prysmaticlabs/prysm/v3/proto/prysm/v1alpha1"
+	"github.com/prysmaticlabs/prysm/v3/testing/assert"
+	"github.com/prysmaticlabs/prysm/v3/testing/require"
+	"github.com/prysmaticlabs/prysm/v3/time/slots"
+	"github.com/prysmaticlabs/prysm/v3/validator/client/beacon-api/mock"
+)
+
+const subscrimeCommitteeSubnetsTestEndpoint = "/eth/v1/validator/beacon_committee_subscriptions"
```
Comment (rkapka) [inline validator/client/beacon-api/subscribe_committee_subnets_test.go:21]: `subscrime` -> `subscribe`

**Review by rkapka — CHANGES_REQUESTED** [review]

File: validator/client/beacon-api/subscribe_committee_subnets.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+		subscribeSlot := in.Slots[index]
+		subscribeCommitteeId := in.CommitteeIds[index]
+		subscribeIsAggregator := in.IsAggregator[index]
+		subscribeValidatorIndex := validatorIndices[index]
+
+		committeesAtSlot, foundSlot := slotToCommitteesAtSlotMap[subscribeSlot]
+		if !foundSlot {
+			// Lazily fetch the committeesAtSlot from the beacon node if they are not already in the map
+			epoch := slots.ToEpoch(subscribeSlot)
+			duties, err := c.dutiesProvider.GetAttesterDuties(ctx, epoch, validatorIndices)
```
[the remark this answers is not in the document]
Comment (PatriceVignola) [reply validator/client/beacon-api/subscribe_committee_subnets.go:37]: Below on line 53, we loop through all slots obtained for the epoch and add them to the map, so essentially we never query an epoch more than once. Even though duties are queried from a specific epoch, committees are slot-based so it simplifies the logic to have a `slot -> committee` map.

**Review by rkapka — APPROVED** [review]

**Review by james-prysm — APPROVED** [review]

**Pull request merged**